File: Backend/app/api/v1/routers/categories.py
from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session

from app.dependencies.auth import get_current_user
from app.dependencies.database import get_db
from app.models.user import User
from app.schemas.category import CategoryCreate, CategoryResponse, CategoryUpdate
from app.services.category_service import CategoryService

router = APIRouter(prefix="/categories", tags=["categories"])


@router.post("", response_model=CategoryResponse, status_code=status.HTTP_201_CREATED)
def create_category(
    data: CategoryCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return CategoryService(db).create_category(current_user.id, data)


@router.get("", response_model=list[CategoryResponse])
def get_categories(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return CategoryService(db).get_all_categories(current_user.id)


@router.get("/{category_id}", response_model=CategoryResponse)
def get_category(
    category_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return CategoryService(db).get_category(category_id, current_user.id)


@router.patch("/{category_id}", response_model=CategoryResponse)
def update_category(
    category_id: int,
    data: CategoryUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return CategoryService(db).update_category(category_id, current_user.id, data)


@router.delete("/{category_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_category(
    category_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    CategoryService(db).delete_category(category_id, current_user.id)

File: Frontend/taskflow/src/api/categories.js
import api from './client'

const CATEGORY_COLORS = [
  '#7C3AED',
  '#0D9488',
  '#F97316',
  '#2563EB',
  '#DB2777',
  '#16A34A',
]

function getCategoryColor(categoryId) {
  return CATEGORY_COLORS[
    categoryId % CATEGORY_COLORS.length
  ]
}

export async function getCategories() {
  const response = await api.get('/categories')

  return response.data.map((category) => ({
    ...category,
    color: getCategoryColor(category.id),
  }))
}

export async function createCategory(name) {
  const response = await api.post('/categories', {
    name,
  })

  return response.data
}

export async function updateCategory(categoryId, name) {
  const response = await api.patch(
    `/categories/${categoryId}`,
    {
      name,
    }
  )

  return response.data
}

export async function deleteCategory(categoryId) {
  await api.delete(`/categories/${categoryId}`)
}